File: tests/src/contracts/project.ts
import { SecretNetworkClient, fromUtf8, Permit, Tx} from "secretjs";
import { entropy } from "../utils";
import { ContractInstance } from "./contracts";

export type ProjectInitMsg = {
    creator: string;
    title: string;
    subtitle?: string;
    description: string;
    pledged_message?: string;
    funded_message?: string;
    reward_messages: RewardMessage[];
    goal: string;
    deadline: number;
    deadman: number;
    categories: number[];
    entropy: string;
    source_contract: string;
    source_hash: string;
    snip20_contract: string;
    snip20_hash: string;
    minimum_pledge: string;
    maximum_pledge: string;
    padding?: string;
}

export type Snip24RewardInit = {
    reward_snip24_code_id: number;
    reward_snip24_code_hash: string;
    name: string;
    admin: string;
    symbol: string;
    decimals: number;
    public_total_supply: boolean;
    enable_deposit: boolean;
    enable_redeem: boolean;
    enable_mint: boolean;
    enable_burn: boolean;
    amount: string;
    contributors_vesting_schedule: VestingEvent[];
    contributors_per_mille: number;
    minimum_contribution: string;
    contribution_weight: number;
    creator_vesting_schedule: VestingEvent[];
    creator_per_mille: number;
    creator_addresses?: string[];
}

export type VestingEvent = {
    block: number;
    per_mille: number;
}

export type ProjectChangeTextMsg = {
    title?: string;
    subtitle?: string;
    description?: string;
    pledged_message?: string;
    funded_message?: string;
    categories?: number[];
    padding?: string;
}

export type ProjectStatusResult = {
    creator: string;
    status: string;
    paid_out: boolean;
    goal: string;
    total: string;
    deadline: number;
    deadman: number;
    title: string;
    subtitle: string;
    description: string;
    categories: number[];
    spam_count: number;
    snip20_address: string,
    minimum_pledge: string,
    maximum_pledge: string,
    pledged_message?: string;
    funded_message?: string;
    reward_messages?: RewardMessage[];
    contribution?: string;
}

export type RewardMessage = {
    threshold: string;
    message: string;
}

export class ProjectContractInstance extends ContractInstance {
    async instantiate(secretjs: SecretNetworkClient, initMsg: ProjectInitMsg, label: string, gasLimit: number = 150_000): Promise<string> {
        return super.instantiate(secretjs, initMsg, label, gasLimit);
    }

    async changeText(secretjs: SecretNetworkClient, changeTextMsg: ProjectChangeTextMsg, gasLimit: number = 150_000): Promise<Tx> {
        const msg = { change_text: changeTextMsg };
        const tx = await this.exec(secretjs, msg, gasLimit);
        return tx;
    }

    async refund(secretjs: SecretNetworkClient, gasLimit: number = 150_000): Promise<Tx> {
        const msg = { refund: { padding: "=========" } };
        const tx = await this.exec(secretjs, msg, gasLimit);
        return tx;
    }

    async cancel(secretjs: SecretNetworkClient, gasLimit: number = 150_000): Promise<Tx> {
        const msg = { cancel: { padding: "=========" } };
        const tx = await this.exec(secretjs, msg, gasLimit);
        return tx;
    }

    async comment(secretjs: SecretNetworkClient, comment: string, gasLimit: number = 150_000): Promise<Tx> {
        const msg = { comment: { comment, padding: "=========" }};
        const tx = await this.exec(secretjs, msg, gasLimit);
        return tx;
    }

    async flag_spam(secretjs: SecretNetworkClient, flag: boolean, gasLimit: number = 150_000): Promise<Tx> {
        const msg = { flag_spam: { flag, padding: "=========" }};
        const tx = await this.exec(secretjs, msg, gasLimit);
        return tx;
    }

    async payOut(secretjs: SecretNetworkClient, gasLimit: number = 150_000): Promise<Tx> {
        const msg = { pay_out: { padding: "=========" } };
        const tx = await this.exec(secretjs, msg, gasLimit);
        return tx;
    }

    async generateViewingKey(secretjs: SecretNetworkClient, gasLimit: number = 150_000): Promise<Tx> {
        const msg = { generate_viewing_key: { entropy: entropy(), padding: "=========" } };
        const tx = await this.exec(secretjs, msg, gasLimit);
        return tx;
    }

    async queryStatus(secretjs: SecretNetworkClient): Promise<ProjectStatusResult> {
        const query = { status: { } };
        const result = (await this.query(secretjs, query)) as ProjectStatusResult;
        return result;
    }

    async queryStatusPermit(secretjs: SecretNetworkClient, permit: Permit): Promise<ProjectStatusResult> {
        const query = { status_with_permit: { permit } };
        const result = (await this.query(secretjs, query)) as ProjectStatusResult;
        return result;
    }

}

File: contracts/mage-hands-platform/src/project.rs

use cosmwasm_std::{Addr, Uint128};
use secret_toolkit::utils::InitCallback;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub struct ProjectInstantiateMsg {
    pub creator: Addr,
    pub title: String,
    pub subtitle: Option<String>,
    pub description: String,
    pub cover_img: String,
    // message for people who have pledged money before funding has been completed
    pub pledged_message: Option<String>,
    // message for people who have contributed money after project is successfully funded
    pub funded_message: Option<String>,
    // reward messages based on contribution thresholds
    pub reward_messages: Vec<RewardMessage>,
    pub goal: Uint128,
    pub deadline: u64,
    pub deadman: u64,
    pub categories: Vec<u16>,

    pub entropy: String,

    pub source_contract: Addr,
    pub source_hash: String,

    pub snip20_contract: Addr,
    pub snip20_hash: String,
    // minimum and maximum pledge amounts
    pub minimum_pledge: Uint128,
    pub maximum_pledge: Uint128,

    // new snip24
    pub snip24_reward_init: Option<Snip24RewardInit>,

    pub padding: Option<String>,
}

#[derive(Serialize, Deserialize, Clone, PartialEq, Debug,)]
pub struct RewardMessage {
    pub threshold: Uint128,
    pub message: String,
}

#[derive(Serialize, Deserialize, Clone, PartialEq, Debug,)]
pub struct Snip24RewardInit {
    pub reward_snip24_code_id: u64,
    pub reward_snip24_code_hash: String,

    // snip24 params
    pub name: String,
    pub admin: Option<Addr>,
    pub symbol: String,
    pub decimals: u8,
    pub public_total_supply: bool,
    pub enable_deposit: bool,
    pub enable_redeem: bool,
    pub enable_mint: bool,
    pub enable_burn: bool,

    // timeline of release of tokens to contributors
    pub contributor_vesting_schedule: Vec<VestingEvent>,
    // minimum contribution to be eligible for reward
    pub minimum_contribution: Option<Uint128>,
    // maximum amount of contribution applied to reward
    // if maximum == 0 or < minimum, then there is no maximum
    pub maximum_contribution: Option<Uint128>,
    // contribution weighting: one of linear, sqrt, or log
    // weighting is applied after minimum and maximum is applied
    pub contribution_weight: u8,

    // timeline of release of tokens to creator
    pub creator_vesting_schedule: Vec<VestingEvent>,
    // addresses to evenly distribute coins to
    // if none, will distribute to project creator address
    pub creator_addresses: Option<Vec<Addr>>,
}

// Vesting events indicate what block
#[derive(Serialize, Deserialize, Clone, PartialEq, Debug,)]
pub struct VestingEvent {
    pub block: u64,
    pub amount: Uint128,
}

impl InitCallback for ProjectInstantiateMsg {
    const BLOCK_SIZE: usize = 256;
}